server/lib/routes/otp.js

```diff
@@ -4,88 +4,116 @@ const uniqid = require('uniqid')
+    return reply.notFound('Subscription not found')
+  }
+
+  const uniqueId = uniqid()
+
+  return new Promise(() => {
+    const requestObj = db.collection('requests').doc(uniqueId)
+    requestObj.set({ createdAt: new Date() })
+
+    const completeRequest = (error, otp) => {
+      requestObj.delete()
+      if (error) {
+        log.error(error.message)
+        return reply.internalServerError()
+      }
 
-      const tokenData = await db.collection('tokens').doc(token).get()
+      unsubscribe()
+      clearTimeout(timeout)
 
-      if (!tokenData.exists) {
-        log.error('Token not found in tokens collection')
-        return reply.notFound('Token not found')
+      if (!otp) {
+        log.error('Request was not approved')
+        return reply.forbidden('Request was not approved')
       }
 
-      const { secretId, subscriptionId } = tokenData.data()
-
-      const subscription = await db
-        .collection('subscriptions')
-        .doc(subscriptionId)
-        .get()
+      log.info('Request approved, sending back OTP')
+      return reply.send(otp)
+    }
 
-      if (!subscription.exists) {
-        log.error(`Subscription not found - ${subscriptionId}`)
-        return reply.notFound('Subscription not found')
+    const unsubscribe = requestObj.onSnapshot(
+      (update) => {
+        const approved = update.get('approved')
+        if (approved === undefined) {
+          // update due to object creation
+          return
+        }
+        completeRequest(null, update.get('otp'))
+      },
+      (error) => {
+        log.error(error.message)
+        return reply.internalServerError()
       }
+    )
 
-      const uniqueId = uniqid()
+    const { name, version } = body.packageInfo
 
-      return new Promise(() => {
-        const requestObj = db.collection('requests').doc(uniqueId)
-        requestObj.set({ createdAt: new Date() })
+    const notification = {
+      subscription: subscription.data(),
+      secretId,
+      uniqueId,
+      token
+    }
+
+    const hasPackageInfo = !!name || !!version
 
-        const completeRequest = (error, otp) => {
-          requestObj.delete()
-          if (error) {
-            log.error(error.message)
-            return reply.internalServerError()
-          }
+    if (hasPackageInfo) {
+      notification.packageInfo = {
+        ...(version && { version }),
+        ...(name && { name })
+      }
+    }
 
-          unsubscribe()
-          clearTimeout(timeout)
+    push.send(notification)
 
-          if (!otp) {
-            log.error('Request was not approved')
-            return reply.forbidden('Request was not approved')
-          }
+    const timeout = setTimeout(completeRequest, approvalLimit)
+  })
+}
 
-          log.info('Request approved, sending back OTP')
-          return reply.send(otp)
-        }
+async function otpRoutes(server) {
+  server.route({
+    method: 'POST',
+    url: '/api/generate/:token',
+    handler: generateOtpHandler(server)
```

Comment:
this pattern is very unusual and I'm not sure I understand why we're doing it this way, so unless there's a valid reason, which unfortunately I could not spot (because the change is quite difficult to review), I think it would be best to stick to a more conventional handler definition pattern instead.

Also, the change in this file is quite hard to review, can you do it in a way that's more obvious? As I assume the actual change is relatively minor, so I don't see a good reason why it should be so hard to review

Reply:
yeah, the diff got a little bit weird, but the gist of it is that I got [this function](https://github.com/nearform/optic/blob/master/server/lib/routes/otp.js#L11), moved it out of the route definition and named it 'generateOtpHandler' so it can be used on both the GET and the POST requests.


I just realised while reading the docs that I can actually use an array of methods and register both with the same handler easily.

Reply:
just fixed that, should be way better now.